src/promptflow/promptflow/_core/openai_injector.py

```diff
@@ -81,50 +121,103 @@ def get_aoai_telemetry_headers() -> dict:
+
+        @functools.wraps(f)
+        async def wrapper(*args, **kwargs):
+            inject_headers(kwargs)
+            return await f(*args, **kwargs)
+
+    else:
+
+        @functools.wraps(f)
+        def wrapper(*args, **kwargs):
+            inject_headers(kwargs)
+            return f(*args, **kwargs)
 
     return wrapper
 
 
-def inject(f):
-    wrapper_fun = inject_operation_headers((inject_function(["api_key", "headers"])(f)))
+def inject_async(f):
+    wrapper_fun = inject_operation_headers((inject_function_async(["api_key", "headers", "extra_headers"])(f)))
+    wrapper_fun._original = f
+    return wrapper_fun
+
+
+def inject_sync(f):
+    wrapper_fun = inject_operation_headers((inject_function_sync(["api_key", "headers", "extra_headers"])(f)))
     wrapper_fun._original = f
     return wrapper_fun
 
 
-def available_openai_apis():
+def _openai_api_list(sync=True):
     if IS_LEGACY_OPENAI:
-        for api in ("Completion", "ChatCompletion", "Embedding"):
-            try:
-                openai_api = getattr(openai, api)
-                if hasattr(openai_api, "create"):
-                    yield openai_api
-            except AttributeError:
-                # This is expected for older versions of openai or unsupported APIs.
-                # E.g. ChatCompletion API was introduced in 2023 and requires openai>=0.27.0 to work.
-                # Older versions of openai do not have this API and will raise an AttributeError if we try to use it.
-                pass
+        sync_apis = (
+            ("openai", "Completion", "create"),
+            ("openai", "ChatCompletion", "create"),
+            ("openai", "Embedding", "create"),
+        )
+
+        async_apis = (
+            ("openai", "Completion", "acreate"),
+            ("openai", "ChatCompletion", "acreate"),
+            ("openai", "Embedding", "acreate"),
+        )
     else:
-        for api in ("Completions", "Chat", "Embeddings"):
-            try:
-                if api == "Chat":
-                    openai_api = getattr(openai.resources.chat, "Completions")
-                else:
-                    openai_api = getattr(openai.resources, api)
-                if hasattr(openai_api, "create"):
-                    yield openai_api
-            except Exception:
-                # To avoid breaking changes included in future upgrades, we ignore all exceptions here.
-                pass
+        sync_apis = (
+            ("openai.resources.chat", "Completions", "create"),
+            ("openai.resources", "Completions", "create"),
+            ("openai.resources", "Embeddings", "create"),
+        )
+
+        async_apis = (
+            ("openai.resources.chat", "AsyncCompletions", "create"),
+            ("openai.resources", "AsyncCompletions", "create"),
+            ("openai.resources", "AsyncEmbeddings", "create"),
+        )
+
+    yield from sync_apis if sync else async_apis
+
+
+def _generate_api_and_injector(apis, sync=True):
```

Comment:
Seems we can do something like this:
```
api_tuples = [(sync_apis, inject_sync), (async_apis, inject_async)]
for apis, injector in api_tuples:
    for module_name, class_name, method_name in apis:
         # Do stuffs
```

Reply:
What's the benefit of doing this?

Reply:
I saw a lot of `sync=True`, I think it is not necessary.

Reply:
If sync and async is mixed, we will have to know which is sync which is async.

Reply:
Updated. Please help review.